generator/_render.py

```diff
@@ -0,0 +1,288 @@
+                ),
+            },
+        }
+    """
+    template = get_template('wrappers.md')
+    outfile = os.path.join(out, 'wrappers.md')
+    template.stream().dump(outfile)
+    with open(outfile) as stream:
+        content = stream.read()
+    wrappers_info = {}
+    method = collections.namedtuple('Method', 'name link doc')
+    for part in content.split('\n## '):
+        if part.startswith('#'):
+            continue
+        facade_name, method_doc = part.split('.', 1)
+        method_name = method_doc.split('(', 1)[0]
+        method_link = method_doc.split('\n', 1)[0].replace(' ', '-')
+        for char in '(),':
+            method_link = method_link.replace(char, '')
+        methods = wrappers_info.setdefault(
+            facade_name, collections.OrderedDict())
+        methods[method_name] = method(
+            name=method_name, link=method_link, doc='## ' + method_doc)
+    return wrappers_info
+
+
+def _handle_methods(name, methods, types):
+    if not isinstance(methods, list):
+        raise TypeError('methods for facade {} is not a list'.format(name))
+    resulting_methods = []
+    for method in methods:
+        if not isinstance(method, collections.Mapping):
+            raise TypeError('method for facade {} is not a map'.format(name))
+        try:
+            method_name = method['Name']
+        except KeyError:
+            raise ValueError('method for facade {} has no name'.format(name))
+        try:
+            params = from_gotype(types, method.get('Param'))
+            result = from_gotype(types, method.get('Result'))
+        except (KeyError, ValueError) as err:
+            raise ValueError(
+                'method {}.{} is not valid: {}'.format(name, method_name, err))
+        doc = method.get('Doc', 'There is no documentation for this method.')
+        resulting_methods.append(Method(
+            request=method_name,
+            params=params,
+            result=result,
+            doc=wrap_text(doc, indent='    ', trailing_newline=True)))
+    # Resulting methods are sorted mostly for making tests deterministic.
+    return sorted(resulting_methods)
+
+
+def _hyphenize(string):
+    """Convert thisString into this-string."""
+    parts = []
+    for char in string:
+        if char.isupper():
+            char = '-' + char.lower()
+        parts.append(char)
+    return ''.join(parts)
+
+
+def _nowstr():
+    """Return the current time in UTC as a string."""
+    return datetime.utcnow().strftime('%a %Y/%m/%d %H:%M:%S UTC')
+
+
+# A black list for facades that surely are not available to users.
+_BLACKLIST = (
+    'AgentTools',
+    'EntityWatcher',
+    'FanConfigurer',
+    'FilesystemAttachmentsWatcher',
+    'Firewaller',
+    'InstancePoller',
+    'MigrationFlag',
+    'MigrationMaster',
+    'MigrationMinion',
+    'MigrationStatusWatcher',
+    'MigrationTarget',
+    'ProxyUpdater',
+    'StorageProvisioner',
+    'StringsWatcher',
+    'Undertaker',
```

Comment:
That's a grim facade name. :stuck_out_tongue_closed_eyes:

Reply:
Lol yes indeed. You made me curious: apparently this is an API for managing a worker which processes a dying model: https://github.com/juju/juju/blob/develop/worker/undertaker/undertaker.go#L53